Feature/29

chrisbrantley opened the pull request:
Fixes issue #29

efryvc commented on drf_jsonapi/mixins.py:
```diff
@@ -45,7 +44,8 @@ def list(self, request, *args, **kwargs):
             only_fields=request.fields,
             include=request.include,
-            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE)
+            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE),
```
Since we're passing the request now, does it make sense to just extract the page size on the other side (in the serializer_class)? Also, why does this line reference `self.request` when we're passing in `request` as an argument to this method. Why do we have `request` and `self.request` in the same method?

chrisbrantley replied:
I think it makes sense to keep the fields, includes, and page_size discreet arguments so they *COULD* be independent of the request. We are passing the request now to allow for user-based filtering and validation of relationships (this follows the preferred pattern of DRF).

To your question about `self.request` vs `request`: I agree it should be `request`. I'll make that change. Django passes in the request AND sets it on the view instance so either would work but it should be consistent and using the value passed in is easier to test. Thanks for catching that!

efryvc commented on drf_jsonapi/mixins.py:
```diff
@@ -118,7 +118,10 @@ def process_relationships(self, relationship_data, resource):
 
-            handler.set_related(resource, related_resources)
+            try:
+                handler.set_related(resource, related_resources, request)
```
I don't love this try/except here. Can we make `set_related` deal with the case where there's no `request` argument, or no?

morgandenning replied:
@chrisbrantley Late Late Reply, but...


Is this something that we can add a deprecation warning (https://docs.python.org/3/library/exceptions.html#DeprecationWarning) on, so that in the next (first?) major version, it can be removed while giving end users a heads-up? I don't mind adding a warning on a later PR, just an idea.

Since this repo is OSS, it'd be nice to follow some common standards such as this.

chrisbrantley replied:
@morgandenning Excellent suggestion. I just pushed a commit that adds those deprecation warnings.

efryvc commented on drf_jsonapi/mixins.py:
```diff
@@ -142,7 +145,8 @@ def create(self, request):
             only_fields=request.fields,
             include=request.include,
-            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE)
+            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE),
```
As above, why are we referencing `self.request` when we're passing `request` into this function as an argument?

chrisbrantley replied:
I'll update these as well.

efryvc commented on drf_jsonapi/mixins.py:
```diff
@@ -154,12 +158,12 @@ def create(self, request):
         # Check for relationships and process them
         if 'relationships' in request.data['data']:
-            resource = self.process_to_one_relationships(request.data['data']['relationships'], resource)
+            resource = self.process_to_one_relationships(request.data['data']['relationships'], resource, request)
```
do we need to pass `request.data` as an arument separately from `request`, or can we have the called function extract the data it needs?

chrisbrantley replied:
I can see arguments for either side. I left it this way to make it backwards compatible and to make testing easier. The `request` is not STRICTLY required but the request data and the resource are. Passing the request here allows the implementation to use it to make user-specific decisions.

efryvc commented on drf_jsonapi/mixins.py:
```diff
@@ -188,7 +192,8 @@ def retrieve(self, request, *args, **kwargs):
             only_fields=request.fields,
             include=request.include,
-            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE)
+            page_size=self.request.GET.get('page[size]', settings.DEFAULT_PAGE_SIZE),
```
Again, why are we referencing `self.request`, when we're passing `request` as an argument?

Also, do we need to pass an element of the request (`page[size]`) when we're passing the request in as `context`?

chrisbrantley replied:
I will remove the `self.`

`page_size` is a discreet property that we are pulling from the request but we could just as easily hard-code it or specify some other way. By only passing the request we are leaving the serializer to have to figure out how to get it from the request (not the serializer's job).

chrisbrantley commented:
@efryvc Tests are under the rug in the upstairs bedroom. You didn't see them?

(I'll write tests and re-submit)

chrisbrantley commented:
I think all the issues on this have been addressed. I fixed a couple of the issues @efryvc pointed out and added a bunch more tests to get the coverage back up. There was a question about why are we passing discreet values that could be extracted from the request (now that we are passing the request) but I hope I've addressed why that's important. If there is still some confusion or disagreement I'm happy to chat about it.

chrisbrantley commented:
I like it. I’ll add the deprecation warnings.

On Thu, Oct 18, 2018 at 12:15 PM Morgan Denning <notifications@github.com>
wrote:

> *@morgandenning* commented on this pull request.
> ------------------------------
>
> In drf_jsonapi/mixins.py
> <https://github.com/Vacasa/drf-jsonapi/pull/32#discussion_r226430532>:
>
> > @@ -118,7 +118,10 @@ def process_relationships(self, relationship_data, resource):
>                  data, many=handler.many
>              )
>
> -            handler.set_related(resource, related_resources)
> +            try:
> +                handler.set_related(resource, related_resources, request)
>
> @chrisbrantley <https://github.com/chrisbrantley> Late Late Reply, but...
>
> Is this something that we can add a deprecation warning (
> https://docs.python.org/3/library/exceptions.html#DeprecationWarning) on,
> so that in the next (first?) major version, it can be removed while giving
> end users a heads-up? I don't mind adding a warning on a later PR, just an
> idea.
>
> Since this repo is OSS, it'd be nice to follow some common standards such
> as this.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Vacasa/drf-jsonapi/pull/32#discussion_r226430532>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB-n_jR4jQJDoequSKmtyvcCGVKW6WCzks5umNNNgaJpZM4WhzI2>
> .
>

chrisbrantley commented:
@efryvc @morgandenning Can this be approved.

chrisbrantley commented:
@efryvc @morgandenning conflict has been resolved and tests fixed. Can we get this approved and merged?

efryvc commented:
I made a couple changes on here because I thought it would be quicker than
commenting. I have approved, but @Chris Brantley
<chris.brantley@vacasa.com> you
should probably give it a once-over.

On Sat, Nov 24, 2018 at 10:47 AM Chris Brantley <notifications@github.com>
wrote:

> @efryvc <https://github.com/efryvc> @morgandenning
> <https://github.com/morgandenning> conflict has been resolved and tests
> fixed. Can we get this approved and merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Vacasa/drf-jsonapi/pull/32#issuecomment-441387800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhmiKmMVUFFpOfT0w9tmrRjNySSUflw2ks5uyZQngaJpZM4WhzI2>
> .
>


-- 



<http://vacasa.com/>

Ethan Fry | Architect

m: 510.599.9298

vacasa.com

chrisbrantley commented:
@efryvc Changes look good. Thank you!
